design-docs/improved-console.md

```diff
@@ -0,0 +1,422 @@
+ - `void update(ProgressOperations operations)`
+ - `void setStatusBarFormatter(StatusBarFormatter formatter)` 
+
+> NOTE: Plugins can provide progress indication by publishing `BuildOperations` (separate effort)
+
+Implementation is similar to `ConsoleBackedProgressRenderer` with throttling and a data
+structure for storing recent updates.
+
+Warnings and errors are displayed in the main area with the same format and styling as today.
+
+`FixedSizeMultiStatusTextArea` would be an implementation of `MultiStatusTextArea` that
+initializes a `List<LinkedList<ProgressOperation>>` of fixed size (defaults to `org.gradle.workers.max`
+ under `--parallel` or 1 otherwise). It would keep track of `ProgressOperation` chains 
+in-progress to prevent having more in-progress operations than max workers displayed.
+
+> NOTE: We must avoid use of save/restore cursor position unless we switch to a 
+[terminfo](https://en.wikipedia.org/wiki/Terminfo)-based Console handling instead of using ANSI.
+    
+### Test Coverage
+* A terminal size with fewer than max parallel operations displays (rows - 2) operations. 
+ That is, the progressArea never grows taller than the console height - 2 rows.
+* Operation status lines are trimmed at (cols - 1)
+* `System.in` and SystemConsole I/O happens on the mainArea, which is unaffected by other areas
+* Renders well on default macOS Terminal, default Windows Console, Cygwin, PowerShell and
+ common Linux Terminals
+* Console shows idle workers under `--continuous` build
+
+## Story: Log messages can be configured to give more context
+With the removal of task execution logging in the main area, users need to get some additional
+context when warnings or errors occur. 
+
+### User Visible Changes
+Users see log level and build operation description for WARN and ERROR level 
+when rendering log events that have a severity and build operation specified.
+
+**Before**:
+```
+Note: Some input files use unchecked or unsafe operations.
+Note: Recompile with -Xlint:unchecked for details.
+Log message from build logic
+```
+
+**After**:
+```
+[:util:compileJava] Note: Some input files use unchecked or unsafe operations.
+[:util:compileJava] Note: Recompile with -Xlint:unchecked for details.
+Log message from build logic
+```
+
+### Implementation
+The concept of a source operation is added to `RenderableOutputEvent`. `LogEvent`s
+and `StyledTextOutputEvent` can be constructed with a String that generally 
+contains the `BuildOperation` description. 
+
+### Test Cases
+* Log messages without source operation configured renders same as usual
+
+### Open Issues
+* Should categories be leveraged instead?
+* Concept of log level or severity
+* Additions needed to `org.gradle.api.Logger` and `StyledTextOutput` for use in build scripts
+
+## Story: Allow user to include additional info in log messages
+Allow users to specify whether timestamp, log level, and/or category are logged with each message.
```

Comment:
> though I'm not sure this is strictly necessary.

100% a solution to the use case is necessary before we remove the task names. There's no way we should ask our users to detangle the error output of 8 different broken JavaCompile tasks where is all jumbled together and interleaved.

The proposed solution (Allow users to specify whether timestamp, log level, and/or category are logged with each message) doesn't really solve the problem.

Instead, I would do something like group the log output of a build operation (however that came to be) together and include either a header or a prefix. I would do this as a built-in capability and not something the task author or build author or build operator has to take action to enable.